packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -81,194 +109,271 @@ ModalBody.propTypes = {
+    const [prevOpen, setPrevOpen] = useState<boolean>(!!open);
+    const innerModal = useRef<HTMLDivElement>(null);
+    const button = useRef<HTMLButtonElement>(null);
+    const startSentinel = useRef<HTMLButtonElement>(null);
+    const endSentinel = useRef<HTMLButtonElement>(null);
+
+    if (open !== prevOpen) {
+      setisOpen(!!open);
+      setPrevOpen(!!open);
     }
-  }
 
-  function handleBlur({
-    target: oldActiveNode,
-    relatedTarget: currentActiveNode,
-  }) {
-    if (open && currentActiveNode && oldActiveNode) {
-      const { current: bodyNode } = innerModal;
-      const { current: startSentinelNode } = startSentinel;
-      const { current: endSentinelNode } = endSentinel;
-      wrapFocus({
-        bodyNode,
-        startSentinelNode,
-        endSentinelNode,
-        currentActiveNode,
-        oldActiveNode,
-        selectorsFloatingMenus,
-      });
+    function handleKeyDown(evt: KeyboardEvent) {
+      if (evt.key === 'Escape') {
+        closeModal(evt);
+      }
+
+      onKeyDown(evt);
     }
-  }
 
-  function closeModal(evt) {
-    if (!onClose || onClose(evt) !== false) {
-      setisOpen(false);
+    function handleMousedown(evt: MouseEvent) {
+      if (
+        !innerModal.current?.contains(evt.target as Node) &&
+        preventCloseOnClickOutside
+      ) {
+        return;
+      }
+      if (innerModal.current?.contains(evt.target as Node)) {
+        closeModal(evt);
+      }
     }
-  }
 
-  const modalClass = cx({
-    [`${prefix}--modal`]: true,
-    'is-visible': isOpen,
-    [customClassName]: customClassName,
-    [`${prefix}--modal--danger`]: danger,
-  });
-
-  const containerClass = cx({
-    [`${prefix}--modal-container`]: true,
-    [`${prefix}--modal-container--${size}`]: size,
-    [`${prefix}--modal-container--full-width`]: isFullWidth,
-    [containerClassName]: containerClassName,
-  });
-
-  // Generate aria-label based on Modal Header label if one is not provided (L253)
-  let generatedAriaLabel;
-  const childrenWithProps = React.Children.toArray(children).map((child) => {
-    switch (child.type) {
-      case React.createElement(ModalHeader).type:
-        generatedAriaLabel = child.props.label;
-        return React.cloneElement(child, {
-          closeModal: closeModal,
-        });
-      case React.createElement(ModalFooter).type:
-        return React.cloneElement(child, {
-          closeModal: closeModal,
-          inputref: button,
+    function handleBlur({
+      target: oldActiveNode,
+      relatedTarget: currentActiveNode,
+    }) {
+      if (open && currentActiveNode && oldActiveNode && innerModal.current) {
+        const { current: bodyNode } = innerModal;
+        const { current: startSentinelNode } = startSentinel;
+        const { current: endSentinelNode } = endSentinel;
+        wrapFocus({
+          bodyNode,
+          startTrapNode: startSentinelNode,
+          endTrapNode: endSentinelNode,
```

Comment:
These were apparently invalid keys here & re-mapping seemed like the much easier option